web/app/mu-plugins/filters.php

```diff
@@ -97,17 +101,96 @@ function fix_core_resource_urls( $url ) {
  *
+ * @since 1.1.0
  * @param string $url The URL to check.
- *
  * @return string The filtered URL.
  */
-function adjust_main_site_urls( $url ) {
+function adjust_main_site_urls( string $url ) : string {
 	// If this is the main site, drop the /wp.
-	if ( is_main_site() ) {
+	if ( is_main_site() && ! __is_login_url( $url ) ) {
 		$url = str_replace( '/wp/', '/', $url );
 	}
 
 	return $url;
 }
-add_filter( 'home_url', 'adjust_main_site_urls', 9 );
-add_filter( 'site_url', 'adjust_main_site_urls', 9 );
+add_filter( 'home_url', __NAMESPACE__ . '\\adjust_main_site_urls', 9 );
+add_filter( 'site_url', __NAMESPACE__ . '\\adjust_main_site_urls', 9 );
+
+/**
+ * Add /wp prefix to all admin and login URLs.
+ * Since /wp is where the core files are installed, this normalizes all non-front-facing urls to use the correct url structure.
+ *
+ * @since 1.1.0
+ * @param string $url The URL to check.
+ * @return string The corrected admin or login URL (or the base url if not an admin or login url).
+ */
+function add_wp_prefix_to_login_and_admin_urls( string $url ) : string {
+	if ( __is_login_url( $url ) ) {
+		if ( strpos( $url, '/wp/' ) === false ) {
+			$url = __normalize_wp_url( preg_replace( '/(\/wp-(login|admin))/', '/wp/$1', $url ) );
+		}
+	}
+
+	return $url;
+}
+add_filter( 'login_url', __NAMESPACE__ . '\\add_wp_prefix_to_login_and_admin_urls', 9 );
+add_filter( 'admin_url', __NAMESPACE__ . '\\add_wp_prefix_to_login_and_admin_urls', 9 );
+
+/**
+ * Check the URL to see if it's either an admin or wp-login URL.
+ *
+ * Validates that the URL is actually a URL before checking.
+ *
+ * @since 1.1.0
+ * @param string $url The URL to check.
+ * @return bool True if the URL is a login or admin URL. False if it's not or is not actually a URL.
+ */
+function __is_login_url( string $url ) : bool {
+	$is_login_url = false;
+
+	// Validate that the string passed was actually a URL.
+	if ( ! preg_match( '/^https?:\/\//i', $url ) ) {
+		$url = 'http://' . ltrim( $url, '/' );
+	}
+
+	// Bail if the string is not a valid URL.
+	if ( ! wp_http_validate_url( $url ) ) {
+		return $is_login_url;
+	}
+
+	// Login page?
+	if ( strpos( $url, 'wp-login' ) !== false ) {
+		$is_login_url = true;
+	}
+
+	// Admin page?
+	if ( strpos( $url, 'wp-admin' ) !== false ) {
+		$is_login_url = true;
+	}
+
+	return $is_login_url;
+}
+
+/**
+ * Remove double-slashes (that aren't http/https://) from URLs.
+ *
+ * @since 1.1.0
+ * @param string $url The URL to test.
+ * @return string The normalized URL.
+ */
+function __normalize_wp_url( string $url ) : string {
+	$scheme_end_pos = strpos( $url, '://' );
+
+	// Extract the scheme from the URL.
+	if ( $scheme_end_pos !== false ) {
+		$scheme = substr( $url, 0, $scheme_end_pos + 3 );
+		$remaining_url = substr( $url, $scheme_end_pos + 3 );
+	} else {
+		$scheme = '';
+		$remaining_url = $url;
+	}
+
+	// Normalize the URL to remove any double slashes.
+	$url = str_replace( '//', '/', $remaining_url );
+	return $scheme . $url;
+}
```

Comment:
Feels dirty not to use parse_url() to get the scheme.
```suggestion
function __normalize_wp_url(string $url): string {
    $scheme = parse_url($url, PHP_URL_SCHEME);

    $scheme_with_separator = $scheme ? $scheme . '://' : '';

    // Remove the scheme from the URL if it exists.
    $remaining_url = $scheme ? substr($url, strlen($scheme_with_separator)) : $url;

    // Normalize the remaining URL to remove any double slashes.
    $normalized_url = str_replace('//', '/', $remaining_url);

    // Reconstruct and return the full normalized URL.
    return $scheme_with_separator . $normalized_url;
}
```
You'll laugh but I wrote this twice because I couldn't decide if I disliked ternary ifs or else statements more, and I settled on allowing ternaries today.